src/main/java/appeng/fluids/parts/PartFluidFormationPlane.java

```diff
@@ -0,0 +1,230 @@
+
+import appeng.api.AEApi;
+import appeng.api.config.AccessRestriction;
+import appeng.api.config.Actionable;
+import appeng.api.config.IncludeExclude;
+import appeng.api.config.Upgrades;
+import appeng.api.networking.events.MENetworkCellArrayUpdate;
+import appeng.api.networking.events.MENetworkChannelsChanged;
+import appeng.api.networking.events.MENetworkEventSubscribe;
+import appeng.api.networking.events.MENetworkPowerStatusChange;
+import appeng.api.networking.security.IActionSource;
+import appeng.api.parts.IPartModel;
+import appeng.api.storage.IMEInventoryHandler;
+import appeng.api.storage.IStorageChannel;
+import appeng.api.storage.channels.IFluidStorageChannel;
+import appeng.api.storage.data.IAEFluidStack;
+import appeng.api.storage.data.IItemList;
+import appeng.api.util.AEPartLocation;
+import appeng.core.sync.GuiBridge;
+import appeng.fluids.util.AEFluidInventory;
+import appeng.fluids.util.IAEFluidInventory;
+import appeng.fluids.util.IAEFluidTank;
+import appeng.items.parts.PartModels;
+import appeng.me.GridAccessException;
+import appeng.me.storage.MEInventoryHandler;
+import appeng.parts.automation.PartAbstractFormationPlane;
+import appeng.parts.automation.PlaneModels;
+import appeng.util.Platform;
+import appeng.util.prioritylist.PrecisePriorityList;
+
+
+public class PartFluidFormationPlane extends PartAbstractFormationPlane<IAEFluidStack> implements IAEFluidInventory
+{
+	private static final PlaneModels MODELS = new PlaneModels( "part/fluid_formation_plane_", "part/fluid_formation_plane_on_" );
+
+	@PartModels
+	public static List<IPartModel> getModels()
+	{
+		return MODELS.getModels();
+	}
+
+	private final MEInventoryHandler<IAEFluidStack> myHandler = new MEInventoryHandler<>( this, AEApi.instance()
+			.storage()
+			.getStorageChannel( IFluidStorageChannel.class ) );
+	private final AEFluidInventory config = new AEFluidInventory( this, 63 );
+
+	public PartFluidFormationPlane( final ItemStack is )
+	{
+		super( is );
+		this.updateHandler();
+	}
+
+	@Override
+	protected void updateHandler()
+	{
+		this.myHandler.setBaseAccess( AccessRestriction.WRITE );
+		this.myHandler.setWhitelist( this.getInstalledUpgrades( Upgrades.INVERTER ) > 0 ? IncludeExclude.BLACKLIST : IncludeExclude.WHITELIST );
+		this.myHandler.setPriority( this.getPriority() );
+
+		final IItemList<IAEFluidStack> priorityList = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createList();
+
+		final int slotsToUse = 18 + this.getInstalledUpgrades( Upgrades.CAPACITY ) * 9;
+		for( int x = 0; x < this.config.getSlots() && x < slotsToUse; x++ )
+		{
+			final IAEFluidStack is = this.config.getFluidInSlot( x );
+			if( is != null )
+			{
+				priorityList.add( is );
+			}
+		}
+		this.myHandler.setPartitionList( new PrecisePriorityList<IAEFluidStack>( priorityList ) );
+
+		try
+		{
+			this.getProxy().getGrid().postEvent( new MENetworkCellArrayUpdate() );
+		}
+		catch( final GridAccessException e )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)